### PR TITLE
Optimize some SegmentedArrayHelper members for JIT

### PR DESCRIPTION
**Pull request by EgorBo**
These properties pop up in perf traces while it is expected that they should be folded to constants. There are two problems here:
1) `AggressiveOptimization` (`(MethodImplOptions)512`) attribute often ruins `static readonly` related optimizations because JIT is forced to compile a method straight to Tier1 so some types might not be statically initialized yet.
Example:
```csharp
class MyProg
{
    static readonly int Size = int.Parse("42");

    [MethodImpl(MethodImplOptions.NoInlining | 
                MethodImplOptions.AggressiveOptimization)]
    static int GetSize() => Size; 

    static void Main()
    {
        for (int i = 0; i < 100; i++)
        {
            GetSize();
            Thread.Sleep(16);
        }
    }
}
```
Codegen for `GetSize` function:
```asm
; Assembly listing for method MyProg:GetSize():int
       4883EC28             sub      rsp, 40
       48B9F045ABE3F87F0000 mov      rcx, 0x7FF8E3AB45F0
       BA06000000           mov      edx, 6
       E86865865F           call     CORINFO_HELP_GETSHARED_NONGCSTATIC_BASE
       8B052EF01C00         mov      eax, dword ptr [(reloc 0x7ff8e3ab462c)]
       4883C428             add      rsp, 40
       C3                   ret
```
Now, if I remove `AggressiveOptimization` and run it again:
```asm
; Assembly listing for method MyProg:GetSize():int
       B82A000000           mov      eax, 42
       C3                   ret 
```
As a bonus, JIT won't waste time on jitting these functions during startup as R2R is expected to be picked up (it doesn't exist for methods with `AggressiveOptimization`).

2) The next issue is a bit more complicated - if your input type is a struct with a shared generic type, e.g.
```csharp
public struct MyStruct<T>
{
    public T t1;
    public T t2;
}
```
Then `ValueTypeSegmentHelper<T>.field` can't be inlined as is (needs dictionary runtime lookup and jit gives up on those). ~~I moved `SegmentShift` and `OffsetMask` to a non generic type since those don't depend on `T`, but I can't do the same for `SegmentSize`.~~

As I workaround we can record which sizes are the most popular and add fast paths probably. Or use a dictionary, cc @jkotas 

Here is a minimal repro:
```csharp
using System.Runtime.CompilerServices;

public struct MyStruct<T>
{
    public T t1;
    public T t2;
}

static class Program
{
    [MethodImpl(MethodImplOptions.NoInlining)]
    static int Foo<T>()
    {
        return SegmentedArrayHelper.GetSegmentSize<T>();
    }

    static void Main()
    {
        for (int i = 0; i < 100; i++)
        {
            Foo<MyStruct<string>>();
            Thread.Sleep(15);
        }
    }
}

internal static class SegmentedArrayHelper
{
    [MethodImpl(MethodImplOptions.AggressiveInlining)]
    internal static int GetSegmentSize<T>()
    {
        if (Unsafe.SizeOf<T>() == Unsafe.SizeOf<object>())
        {
            return ReferenceTypeSegmentHelper.SegmentSize;
        }
        else
        {
            return ValueTypeSegmentSizeHelper<T>.SegmentSize;
        }
    }

    private static class ValueTypeSegmentSizeHelper<T>
    {
        public static readonly int SegmentSize = CalculateSegmentSize(Unsafe.SizeOf<T>());

        private static int CalculateSegmentSize(int elementSize)
        {
            const int Threshold = 85000;
            var segmentSize = 2;
            while (ArraySize(elementSize, segmentSize << 1) < Threshold)
            {
                segmentSize <<= 1;
            }
            return segmentSize;
            static int ArraySize(int elementSize, int segmentSize)
            {
                return (2 * IntPtr.Size) + (elementSize * segmentSize);
            }
        }
    }

    private static class ReferenceTypeSegmentHelper
    {
        public static readonly int SegmentSize = CalculateSegmentSize(IntPtr.Size);

        private static int CalculateSegmentSize(int elementSize)
        {
            const int Threshold = 85000;
            var segmentSize = 2;
            while (ArraySize(elementSize, segmentSize << 1) < Threshold)
            {
                segmentSize <<= 1;
            }
            return segmentSize;
            static int ArraySize(int elementSize, int segmentSize)
            {
                return (2 * IntPtr.Size) + (elementSize * segmentSize);
            }
        }
    }
}
```
Codegen for `Foo`:
```asm
; Assembly listing for method Program:Foo[MyStruct`1[System.__Canon]]():int
G_M32814_IG01:              
       4883EC28             sub      rsp, 40
       48894C2420           mov      qword ptr [rsp+20H], rcx
G_M32814_IG02:              
       488B5138             mov      rdx, qword ptr [rcx+38H]
       488B5210             mov      rdx, qword ptr [rdx+10H]
       4885D2               test     rdx, rdx
       7402                 je       SHORT G_M32814_IG04
G_M32814_IG03:             
       EB12                 jmp      SHORT G_M32814_IG05
G_M32814_IG04:            
       48BA587CEEE3F87F0000 mov      rdx, 0x7FF8E3EE7C58      ; global ptr
       E8B901525F           call     CORINFO_HELP_RUNTIMEHANDLE_METHOD
       488BD0               mov      rdx, rax
G_M32814_IG05:              
       488BCA               mov      rcx, rdx
       FF15CDE25F00         call     [SegmentedArrayHelper:GetSegmentSize[MyStruct`1[System.__Canon]]():int]
       90                   nop      
G_M32814_IG06:              
       4883C428             add      rsp, 40
       C3                   ret
```

### CODE REVIEW

**Comment by EgorBo** [comment]
Is the segment size calculation logic is still relevant? Considering that GC is now Region-based

File: src/Dependencies/Collections/Internal/SegmentedArrayHelper.cs [hunk 1]
```diff
@@ -16,14 +16,7 @@ internal static class SegmentedArrayHelper
...
-        /// <see cref="F:System.Runtime.CompilerServices.MethodImplOptions.AggressiveOptimization"/>.
-        /// </summary>
-        [SuppressMessage("Documentation", "CA1200:Avoid using cref tags with a prefix", Justification = "The field is not supported in all compilation targets.")]
-        internal const MethodImplOptions FastPathMethodImplOptions = MethodImplOptions.AggressiveInlining | (MethodImplOptions)512;
-
-        [MethodImpl(FastPathMethodImplOptions)]
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
         internal static int GetSegmentSize<T>()
         {
             if (Unsafe.SizeOf<T>() == Unsafe.SizeOf<object>())
```
[the remark this answers is not in the document]
Comment (EgorBo) [reply src/Dependencies/Collections/Internal/SegmentedArrayHelper.cs:22]: It complains on `error CS0208: Cannot take the address of, get the size of, or declare a pointer to a managed type ('T')` that I can't suppress 🤔

[the remark this answers is not in the document]
Comment (jkotas) [reply src/Dependencies/Collections/Internal/SegmentedArrayHelper.cs:22]: Is it building with recent Roslyn? The ability to suppress this was added not too long ago.

**Comment by JulieLeeMSFT** [comment]
Fixes https://github.com/dotnet/roslyn/pull/67558.

File: src/Dependencies/Collections/Internal/SegmentedArrayHelper.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (EgorBo) [reply src/Dependencies/Collections/Internal/SegmentedArrayHelper.cs:22]: I assume it builds with the bootstrap SDK which is `8.0.100-preview.1.23115.2`

[the remark this answers is not in the document]
Comment (jkotas) [reply src/Dependencies/Collections/Internal/SegmentedArrayHelper.cs:22]: It is possible to suppress this in .NET 7 SDK. For example, this compiles fine with .NET 7 SDK:

```
internal unsafe static int GetSegmentSize<T>()
{
#pragma warning disable CS8500 // Takes a pointer to a managed type
    return (sizeof(T) == sizeof(object)) ? 1 : -1;
#pragma warning restore CS8500
}
```

[the remark this answers is not in the document]
Comment (EgorBo) [reply src/Dependencies/Collections/Internal/SegmentedArrayHelper.cs:22]: I think I've already tried that in the Roslyn repo and but it still complained during build, I'll try again tomorrow

[the remark this answers is not in the document]
Comment (EgorBo) [reply src/Dependencies/Collections/Internal/SegmentedArrayHelper.cs:22]: Yes, still complains, perhaps someone from Roslyn can help? Although, I think I fixed the main perf issue in this PR already

File: src/Dependencies/Collections/Internal/SegmentedArrayHelper.cs [hunk 2]
```diff
@@ -36,10 +37,11 @@ internal static int GetSegmentShift<T>()
             {
                 return ReferenceTypeSegmentHelper.SegmentShift;
```
Comment (sharwell) [inline src/Dependencies/Collections/Internal/SegmentedArrayHelper.cs:38]: ❔ Any reason to keep `ReferenceTypeSegmentHelper` for .NET execution?

File: src/Dependencies/Collections/Internal/SegmentedArrayHelper.cs [hunk 3]
```diff
@@ -114,6 +117,31 @@ private static int CalculateOffsetMask(int segmentSize)
             return segmentSize - 1;
         }
 
+        // Faster inline implementation for NETCOREAPP to avoid static constructors and non-inlineable
+        // generics with runtime lookups
+#if NETCOREAPP
```
[the remark this answers is not in the document]
Comment (EgorBo) [reply src/Dependencies/Collections/Internal/SegmentedArrayHelper.cs]: Thanks! addressed

**Review by sharwell — APPROVED** [review]

**Comment by sharwell** [comment]
@dotnet/roslyn-compiler I'll submit a follow-up pull request to manually constant-fold the results for the most common item sizes (4, 8, 24, and 40), and also fix the logic as mentioned in https://github.com/dotnet/roslyn/pull/67558#issuecomment-1489441636.

**Review by jaredpar — APPROVED** [review]

**Review by AlekseyTs — APPROVED** [review]
LGTM (commit 7)

File: src/Dependencies/Collections/Internal/SegmentedArrayHelper.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (raulsntos) [reply src/Dependencies/Collections/Internal/SegmentedArrayHelper.cs]: Did you mean `NETCOREAPP3_0_OR_GREATER`?

**Pull request merged**